Path object updates and other fixes

mahesh-panchal opened the pull request:
<!--
# sanger-tol/treeval pull request

Many thanks for contributing to sanger-tol/treeval!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/sanger-tol/treeval/tree/master/.github/CONTRIBUTING.md)
-->

## Changes

- Replace `GrabFiles` process with `resolve`.
- Remove `ifEmpty(null)` injections.
- Convert path strings to file objects as soon as possible.
- Modify `YAML_INPUT` process to emit data directly instead of split and recombine with channels.
- Fix `read_ch` logic in `YAML_INPUT` which uses channels in `if`.


## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/treeval/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

DLBPointon commented:
Morning @mahesh-panchal,

Thanks for the PR, I've changed the base to `dp24_specified_reads` which is the current release candidate for 1.4.0. This RC is the 3.2 template (at this rate it'll be 3.3) and has some substantial changes through the pipeline.

This removes all the version(null) stuff.

I'll have to look at your re-write of YAML_INPUT, i like it but it's just not a way i'm familiar with.

Only note I have is that because of the resolve on busco output for the full table, the code is duplicated in ancestral and as the full busco dir isn't used at all it can be deleted and the resolve of busco full table could simply be passed into ancestral. Hopefully that makes sense. It's where we've tacked on the ancestral workflow without correctly integrating it in the past.

Testing I can do although I won't be able to test on our cluster, as the Sanger HPC is currently dead until next week at least.

mahesh-panchal commented:
> Thanks for the PR, I've changed the base to `dp24_specified_reads` which is the current release candidate for 1.4.0. This RC is the 3.2 template (at this rate it'll be 3.3) and has some substantial changes through the pipeline.

Thanks. I've rebased against this branch.

> I'll have to look at your re-write of YAML_INPUT, i like it but it's just not a way i'm familiar with.

Sorry about that, but I felt it would be better for maintainability. It's hard to see what's coming out of the channels generally, and there are tests baked into various subsections. For example on one `multiMap` it tested for workflow == "FULL", but then on a multimap that emitted parts of that, it had workflow == "FULL" && workflow == "JBROWSE" ( or something like that ), but the second part would never be true because the input only satisfies the first condition.

It's basically just made one multiMap, and each channel groups and emits the components that are combined with different channels. There are some parts at the top to define some reused variables.

> 
> Only note I have is that because of the resolve on busco output for the full table, the code is duplicated in ancestral and as the full busco dir isn't used at all it can be deleted and the resolve of busco full table could simply be passed into ancestral. Hopefully that makes sense. It's where we've tacked on the ancestral workflow without correctly integrating it in the past.
> 

I'm not sure I understand, but let me take a closer look at the code.

> Testing I can do although I won't be able to test on our cluster, as the Sanger HPC is currently dead until next week at least.
> 
> NOTE: The main difference is template, and longread and hic data are now explicitly input to the pipeline, no more supply a folder and let the pipeline do the work. it has meant that the yaml_input has big chunk of duplicated code. But this will change a fair bit of your work in some of the subworkflows.

I'll investigate. Thank you for letting me know. I'll get to work on fixing the conflicts now.

DLBPointon commented on subworkflows/local/busco_annotation/main.nf:
```diff
@@ -46,7 +46,7 @@ workflow BUSCO_ANNOTATION {
     )
     ch_versions                 = ch_versions.mix(BUSCO_BUSCO.out.versions.first())
-    ch_grab                     = GrabFiles(BUSCO_BUSCO.out.busco_dir)
+    ch_grab                     = BUSCO_BUSCO.out.busco_dir.map { meta, dir -> tuple(meta, files(dir.resolve("*/*/full_table.tsv"), checkIfExists: true)) }
```
1 - `ch_grab` is passed into ANCESTRAL_GENE as `ch_busco_lep_data.busco_dir` as the full_table.tsv now so the new ch_grab made in ANCESTRAL_GENE is no longer needed (as that was used to get the full_table, again).

mahesh-panchal replied:
Sorry, I'm not quite following here. The channel says `busco_dir` but it's not the directory and only the full_table.tsv?

mahesh-panchal commented on subworkflows/local/ancestral_gene/main.nf:
```diff
@@ -18,7 +18,7 @@ workflow ANCESTRAL_GENE {
     ch_versions             = Channel.empty()
 
-    ch_grab                 = GrabFiles(busco_dir)
+    ch_grab                 = busco_dir.map { meta, dir -> tuple(meta, files(dir.resolve("*/*/full_table.tsv"), checkIfExists: true)) }
```
OK. I removed `ch_grab` here, but there's still a reference to it on l.33. I just got lost in the changes now. 
So 
```

DLBPointon commented:
This looks great to me Mahesh, thanks for the work.
I'm getting @weaglesBio to review as well in case there is anything i've missed.
